Add javascript support

mrkutly opened the pull request:
Adds support for javascript as the build target.

**Changes:**
- Upgrades `gleeunit` to `~> 0.6` (there is a bug in the javascript output of the current version that throws errors when running tests with the target set to javascript).
- Adds javascript ffi

lpil commented on src/gleam/json.gleam:
```diff
@@ -127,8 +170,19 @@ pub external fn string(input: String) -> Json =
+
+if javascript {
+  external fn do_bool(Bool) -> Json =
+    "../gleam_json_ffi.mjs" "bool"
```
This can just be the identity function

lpil commented on src/gleam/json.gleam:
```diff
@@ -139,8 +193,19 @@ pub external fn bool(input: Bool) -> Json =
+
+if javascript {
+  external fn do_int(Int) -> Json =
+    "../gleam_json_ffi.mjs" "int"
```
Identity here too

lpil commented on src/gleam/json.gleam:
```diff
@@ -151,8 +216,19 @@ pub external fn int(input: Int) -> Json =
+
+if javascript {
+  external fn do_float(input: Float) -> Json =
+    "../gleam_json_ffi.mjs" "float"
```
And this one

lpil commented on src/gleam/json.gleam:
```diff
@@ -199,8 +286,19 @@ pub fn nullable(from input: Option(a), of inner_type: fn(a) -> Json) -> Json {
+
+if javascript {
+  external fn do_object(entries: List(#(String, Json))) -> Json =
+    "../gleam_json_ffi.mjs" "object_from"
```
Let's call it `object`? To match the Erlang version

lpil commented on src/gleam_json_ffi.mjs:
```diff
@@ -0,0 +1,70 @@
+
+
+export function decode(bit_string) {
+  const stringResult = bit_string_to_string(bit_string)
```
This conversion to a bit string in JS is slow! Let's always operate on the string when on the JS target, to avoid this conversion. It'll have to be the other way round from Erlang

lpil commented on src/gleam_json_ffi.mjs:
```diff
@@ -0,0 +1,70 @@
+  return err.message === 'Unexpected end of JSON input'
+}
+
+const unexpectedByteRegex = /Unexpected token (.) in JSON at position (\d+)/
```
Is this going to work on all JavaScript runtimes?

mrkutly replied:
I think it probably will not. I will need to do more research to figure out how to handle specific runtime errors in JS. This is just a very rough go to see if I could get something working.

lpil commented on src/gleam_json_ffi.mjs:
```diff
@@ -0,0 +1,70 @@
+function getJsonDecodeError(stdErr) {
+  if (isUnexpectedByte(stdErr)) return new toUnexpectedByteError(stdErr)
+  if (isUnexpectedEndOfInput(stdErr)) return new UnexpectedEndOfInput()
+  return undefined
```
This returns `Nil` if nothing matches. Let's default to unexpected byte with some incorrect values for the fields. It would be type safe and wouldn't cause crashes, even if it's not perfect.

lpil reviewed (commented):
Brilliant work! Thank you

I've left some little notes

mrkutly commented:
> Brilliant work! Thank you
> 
> I've left some little notes

Thank you so much for taking a look! I'll go another round on this this week.

mrkutly commented:
@lpil I did more research on the ways unexpected end of input and unexpected byte errors are reported in various Javascript runtimes. Particularly:

- Chromium (Chrome, Node, Edge)
- JavascriptCore (Safari)
- Spidermonkey (Firefox)

What I found is that there is no standard specified by ECMAScript other than that all of these are `SyntaxError`s. There is no error code or any kind of metadata specifying what happened. There is also inconsistent data reported in the error messages. For example, safari will report _**what**_ the unexpected byte was, but not **_where_** it is. Similarly, firefox will report _**where**_ the unexpected byte is, but not _**what**_ it is.

For these cases, I tried to do what I could to resolve all the necessary info to create the `UnexpectedByte` errors. For safari though, it seems to me the only way to resolve which byte was unexpected would be to implement a JSON parser from scratch, which seems to defeat the purpose of using `JSON.parse` at all. So in that case, I just reported the position as `0`.

I also found that sometimes firefox and safari will report `unexpected end of input` errors when chromium reports `unexpected byte` errors. I left a comment in the code about this.

Let me know if you have any advice on better ways to handle this! I tried to find if anyone else tackled this issue online but only found runtime-specific error helpers.

lpil reviewed (commented):
Very cool! Loving the new runtime error handling.

There's a lot of logic in the JS here. I think it would be good to have tests for it, to help with future maintainence. We could import key functions as external fns and check that various error formats parse correctly.

mrkutly commented:
Thanks @lpil! That's an awesome idea! I was wondering about the best way to test these JS internals through gleam 😄 

I'm also going to look into the failing specs with the JS target. I still think it is something funky happening in the `gleam/dynamic` module.

lpil commented:
On JavaScript the decoders for list and for tuple will both be attempting to parse JavaScript arrays, so the error message is likely not quite the same as on the Erlang target.

mrkutly commented:
> On JavaScript the decoders for list and for tuple will both be attempting to parse JavaScript arrays, so the error message is likely not quite the same as on the Erlang target.

Ah that explains it! I'll work on updating the tests to handle both targets.

mrkutly commented:
@lpil I've updated the existing tests to handle both targets and added another test module for testing the js ffi. I only tested the `getJsonDecodeError` function as that is the entry point for the rest of the JS logic. Let me know what you think! And thank you again for all the help!

lpil reviewed (commented):
Really great stuff! I've left 1 question inline

lpil commented on test/gleam_json_js_ffi_test.gleam:
```diff
@@ -0,0 +1,72 @@
+import gleam/json.{DecodeError, UnexpectedByte, UnexpectedEndOfInput}
+import gleeunit/should
+
+if javascript {
+  pub fn main() {
+    gleeunit.main()
+  }
+}
```
I think we want to remove this and instead put the tests in a conditional compilation block

lpil reviewed (approved):
Thank you so much!
